### PR TITLE
Create policy for a container_device_t

### PR DESCRIPTION
**Pull request by rhatdan**
This policy can be used for kubernetes/container plugins which add
devices to containers.

Signed-off-by: Daniel J Walsh <dwalsh@redhat.com>

### CODE REVIEW

**Comment by rhatdan** [comment]
Replaces: https://github.com/containers/container-selinux/pull/177

**Comment by rhatdan** [comment]
@[mregmi](https://github.com/mregmi) lets work together on this one.  Could you try this policy out on your system and tell me what AVCs are created?

**Comment by mregmi** [comment]
Thank you. i will test it out today.

**Comment by mregmi** [comment]
Hi Dan, i am seeing error related to connectto.

`type=AVC msg=audit(1650654069.545:12189): avc:  denied  { connectto } for  pid=2653465 comm="intel_sgx_devic" path="/var/lib/kubelet/device-plugins/kubelet.sock" scontext=system_u:system_r:container_device_t:s0:c310,c693 tcontext=system_u:system_r:container_runtime_t:s0 tclass=unix_stream_socket permissive=0`

audit2allow suggests the following
#============= container_device_t ==============
allow container_device_t container_runtime_t:unix_stream_socket connectto;

**Comment by rhatdan** [comment]
That means that kubernetes is not running with the correct label.  Can you run restorecon on the kublet and see if it gets relabeled to kublet_exec_t, and then restart it and make sure it is running as kublet_t.

That would change the access to container_device_t connectto kublet_t which should be allowed.

**Comment by mregmi** [comment]
yes it was not running with new label. restorecon worked. Thank you.

There are a also couple of things we need for this.
There is also an initcontainer for each plugin and they need access to things in /etc/kubernetes/. Those files are kubernetes_file_t. Do you think we need different labels for initcontainer or we can give access to the above label to container_device_t?

Also one of the plugin we are releasing later this year also needs access to sysfs (files in /sys) which have sysfs_t. the container_device_t would also need access to this. Do we want to add this while we are doing this?

lastly the workloads using these plugins will just need access to /dev (wont need connectto and other things). One way they can work is using the setsebool. do you think it would be better to have a workload specific label with just access to /dev/?

**Comment by rhatdan** [comment]
Can you write this up in a table.  When  you say access to sysfs_t, are you talking read or write?  Also are you mounting in /sys from the host into the container?  /sys is usually namespaced.

Are you saying only the init container needs to talk to kublet?  And not the general continer?

Could we just run with 
container_device_t == container_t + read/write devices?
init == container_device_t + read/write kublet socket?

**Comment by mregmi** [comment]
So there are 3 types of the containers for device plugins:
1) The device plugins (the container_device_t in current patch): 
     container_device_t  = container_t + r/w /dev (device_t) + r/w kubelet socket + r/w some parts of /sys (sysfs_t)
      Like /dev, /sys is also volume mounted from host

2) init container
    container_device_init_t = container_t + kubernetes_file_t (because it needs access to /etc/kubernetes/node-feature-discovery  )

3) the third one is the normal containers with end users application. in most cases they just need /dev/ file access.
  container_t + access to /dev (device_t)

 Thanks.

**Comment by mregmi** [comment]
Yes lets go with these:
1) Device plugins: container_device_plugin_t
2) Init containers: container_device_plugin_init_t
3) workload: container_device_t

The workload (3) will need both read and write to device_t. 
Init container (2) will also need read/write access as it copies some file in that folder 
Thanks.

**Comment by rhatdan** [comment]
How does this look?

container_domain_template(container_device)
allow container_device_t device_node:chr_file rw_chr_file_perms;

container_domain_template(container_device_plugin)
container_kublet_stream_connect(container_device_plugin_t)
allow container_device_plugin_t device_node:chr_file rw_chr_file_perms;

container_domain_template(container_device_plugin_init)
container_kublet_stream_connect(container_device_plugin_init_t)
allow container_device_plugin_init_t device_node:chr_file rw_chr_file_perms;
manage_dirs_pattern(container_device_plugin_init_t, kubernetes_file_t, kubernetes_file_t)
manage_files_pattern(container_device_plugin_init_t, kubernetes_file_t, kubernetes_file_t)
manage_lnk_files_pattern(container_device_plugin_init_t, kubernetes_file_t, kubernetes_file_t)

**Comment by mregmi** [comment]
one small comment.
container_device_plugin_init_t does not need socket communication with kubelet. so we can remove container_kublet_stream_connect(container_device_plugin_init_t)

container_device_plugin_t will also need rw access to /sys so a rule allowing to sysfs_t might be needed.

File: container.fc [hunk 1]
```diff
@@ -5,10 +5,10 @@
 /usr/libexec/docker/docker.*	--	gen_context(system_u:object_r:container_runtime_exec_t,s0)
 /usr/local/libexec/docker/docker.*	--	gen_context(system_u:object_r:container_runtime_exec_t,s0)
 /usr/s?bin/docker.*		--	gen_context(system_u:object_r:container_runtime_exec_t,s0)
-/usr/s?bin/kubelet.*		--	gen_context(system_u:object_r:container_runtime_exec_t,s0)
-/usr/local/s?bin/kubelet.*		--	gen_context(system_u:object_r:container_runtime_exec_t,s0)
-/usr/s?bin/hyperkube.*		--	gen_context(system_u:object_r:container_runtime_exec_t,s0)
-/usr/local/s?bin/hyperkube.*		--	gen_context(system_u:object_r:container_runtime_exec_t,s0)
+/usr/s?bin/kubelet.*		--	gen_context(system_u:object_r:kublet_exec_t,s0)
```
Comment (mythi) [inline container.fc:8]: should this be `kubelet_exec_t`?

**Comment by mythi** [comment]
> one small comment. container_device_plugin_init_t does not need socket communication with kubelet. so we can remove container_kublet_stream_connect(container_device_plugin_init_t)
> 
> container_device_plugin_t will also need rw access to /sys so a rule allowing to sysfs_t might be needed.

with some devices, the `initContainer` is used to write to `/sys` (e.g., to enable `sriov_numvfs` for PCIe device)

**Comment by mregmi** [comment]
@rhatdan I also saw this error when testing the latest policy.

type=AVC msg=audit(1651188022.040:40076): avc:  denied  { read write } for  pid=1185191 comm="cp" name="intel-sgx-epchook" dev="sda4" ino=614477480 scontext=system_u:system_r:container_device_plugin_init_t:s0:c178,c513 tcontext=system_u:object_r:container_file_t:s0:c351,c450 tclass=file permissive=1

type=AVC msg=audit(1651188022.058:40077): avc:  denied  { setattr } for  pid=1185191 comm="cp" name="intel-sgx-epchook" dev="sda4" ino=614477480 scontext=system_u:system_r:container_device_plugin_init_t:s0:c178,c513 tcontext=system_u:object_r:container_file_t:s0:c351,c450 tclass=file permissive=1

audit2allow gives this warning.
#============= container_device_plugin_init_t ==============

#!!!! This avc is a constraint violation.  You would need to modify the attributes of either the source or target types to allow this access.
#Constraint rule:
#       mlsconstrain file { ioctl read lock execute execute_no_trans } ((h1 dom h2 -Fail-)  or (t1 != mcs_constrained_type -Fail-) ); Constraint DENIED
mlsconstrain file { write setattr append unlink link rename } ((h1 dom h2 -Fail-)  or (t1 != mcs_constrained_type -Fail-) ); Constraint DENIED
mlsconstrain file { create relabelto } ((h1 dom h2 -Fail-)  and (l2 eq h2)  or (t1 != mcs_constrained_type -Fail-) ); Constraint DENIED
mlsconstrain file { relabelfrom } ((h1 dom h2 -Fail-)  or (t1 != mcs_constrained_type -Fail-) ); Constraint DENIED

#       Possible cause is the source level (s0:c178,c513) and target level (s0:c351,c450) are different.
allow container_device_plugin_init_t container_file_t:file { read setattr write };

**Comment by rhatdan** [comment]
This means that you ran a container with one label which created content and then attempted to read/write the content from a different.  If the content needs to be shared between multiple containers running in different Pods. then the container_file_t content has to be labeled as s0, not with a private label for the container.

File: container.fc (hunk 1 above)
Comment (rhatdan) [reply container.fc:8]: Fixing.

**Comment by rhatdan** [comment]
`with some devices, the initContainer is used to write to /sys (e.g., to enable sriov_numvfs for PCIe device)`

Added, does it need to set sysctls as well or just this one field.

**Comment by mregmi** [comment]
I think  this line dev_rw_sysfs((container_device_plugin_t)
is also needed for `container_device_plugin_init_t`

I will debug the above container_file_t issue.

File: container.te [hunk 2]
```diff
@@ -1298,3 +1298,67 @@ kernel_mounton_core_if(container_engine_t)
...
+
+# Standard container which needs to be allowed to use any device
+container_domain_template(container_device)
+allow container_device_t device_node:chr_file rw_chr_file_perms;
+
+# Standard container which needs to be allowed to use any device and
+# communicate with kublet
+container_domain_template(container_device_plugin)
+allow container_device_plugin_t device_node:chr_file rw_chr_file_perms;
+dev_rw_sysfs((container_device_plugin_t)
```
Comment (mregmi) [inline container.te:1354]: there is a typo. should be dev_rw_sysfs(container_device_plugin_t)

File: container.te [hunk 3]
```diff
@@ -1298,3 +1298,67 @@ kernel_mounton_core_if(container_engine_t)
...
+container_kublet_stream_connect(container_device_plugin_t)
+
+# Standard container which needs to be allowed to use any device and
+# modify kublet configuration
+container_domain_template(container_device_plugin_init)
+allow container_device_plugin_init_t device_node:chr_file rw_chr_file_perms;
+dev_rw_sysfs((container_device_plugin_t)
```
Comment (mregmi) [inline container.te:1361]: should be dev_rw_sysfs(container_device_plugin_init_t)

File: container.te [hunk 4]
```diff
@@ -1298,3 +1298,67 @@ kernel_mounton_core_if(container_engine_t)
...
+optional_policy(`
+	gen_require(`
+		role unconfined_r;
+	')
+	role unconfined_r types kublet_t;
+	unconfined_domain(kublet_t)
+')
+
+
+type kublet_exec_t;
```
Comment (mregmi) [inline container.te:1314]: in container.fc, it was changed to kubelet_exec_t. so these needs to be changed too. i saw module load error without this change.

**Review by mregmi — COMMENTED** [review]
after a quick fix related to the things i commented. the labels works for the plugins. i think we can merge after fixing those small issues.

**Comment by rhatdan** [comment]
@[mregmi](https://github.com/mregmi) @[Walnux](https://github.com/Walnux) PTAL

**Comment by mregmi** [comment]
@rhatdan only one comment at dev_rw_sysfs lines. second one should be  dev_rw_sysfs(container_device_plugin_init_t).
Other than that it looks good. I tested it on our system.

**Comment by mregmi** [comment]
The latest push does not seem to have any change.

**Comment by mregmi** [comment]
LGTM. Thank you @rhatdan for helping out with this issue.

**Pull request merged**